File: frontend/src/lib/tradesync/reflection.ts
export const EMOTIONS = {
  calm: "平静", confident: "自信", anxious: "焦虑", fearful: "恐惧", greedy: "贪婪",
  frustrated: "挫败", impulsive: "冲动", tired: "疲劳", other: "其他",
} as const;
export const REVIEW_ERRORS = {
  chasing: "追单", early_entry: "提前入场", overtrading: "过度交易", revenge: "报复性交易",
  oversized: "超仓", no_stop: "未设置止损", moving_stop: "随意移动止损", early_exit: "过早退出",
  late_exit: "拖延退出", plan_deviation: "偏离计划", other: "其他",
} as const;
export type Emotion = keyof typeof EMOTIONS;
export type ReviewError = keyof typeof REVIEW_ERRORS;
export interface ReviewReflection {
  recording_basis: "retrospective";
  emotion_before: Emotion | null;
  emotion_after: Emotion | null;
  emotion_notes: string;
  error_assessment: "unassessed" | "none" | "identified";
  errors: ReviewError[];
  primary_error: ReviewError | null;
  conclusion: string;
  next_action: string;
  no_new_action: boolean;
}

export function reflectionComplete(value: ReviewReflection) {
  return !!value.conclusion.trim() && (!!value.next_action.trim() || value.no_new_action);
}

File: backend/app/trades/reflection.py
"""Structured, retrospective self-assessment; never inferred from P&L."""
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

Emotion = Literal["calm", "confident", "anxious", "fearful", "greedy", "frustrated", "impulsive", "tired", "other"]
ErrorCode = Literal["chasing", "early_entry", "overtrading", "revenge", "oversized", "no_stop", "moving_stop", "early_exit", "late_exit", "plan_deviation", "other"]


class Reflection(BaseModel):
    model_config = ConfigDict(extra="forbid")
    recording_basis: Literal["retrospective"] = "retrospective"
    emotion_before: Emotion | None = None
    emotion_after: Emotion | None = None
    emotion_notes: str = Field(default="", max_length=2000)
    error_assessment: Literal["unassessed", "none", "identified"] = "unassessed"
    errors: list[ErrorCode] = Field(default_factory=list, max_length=11)
    primary_error: ErrorCode | None = None
    conclusion: str = Field(default="", max_length=2000)
    next_action: str = Field(default="", max_length=2000)
    no_new_action: bool = Field(default=False, strict=True)

    @field_validator("emotion_notes", "conclusion", "next_action")
    @classmethod
    def trim_text(cls, value):
        return value.strip()

    @field_validator("errors")
    @classmethod
    def unique_errors(cls, values):
        return list(dict.fromkeys(values))

    @model_validator(mode="after")
    def consistent_choices(self):
        if self.error_assessment == "identified" and not self.errors:
            raise ValueError("选择存在错误时，请至少选择一项错误分类")
        if self.error_assessment != "identified" and (self.errors or self.primary_error):
            raise ValueError("未评估或未发现错误时，不能同时选择错误分类")
        if self.primary_error and self.primary_error not in self.errors:
            raise ValueError("主要错误必须属于已选择的错误分类")
        if self.no_new_action and self.next_action:
            raise ValueError("无需新增行动与下次行动不能同时填写")
        return self
